src/ORM/Table.php

```diff
@@ -1601,6 +1601,123 @@ public function deleteAll($conditions)
+     * and page.
+     *
+     * ### Options
+     *
+     * - 'finder' string Defaults to 'all'. The custom finder to use when reading batches.
+     * - 'options' array Defaults to []. The options to pass to the finder when reading batches.
+     * - `atomic` boolean Defaults to true. When true the deletion happens within a transaction.
+     * - 'stopOnFailure' boolean Defaults to true. Stop deleting records when a record fails to be deleted.
+     * - 'batch' integer Defaults to 1. Number of records to read in batches, zero reads all records.
+     * - `checkRules` boolean Defaults to true. Check deletion rules before deleting the record.
+     *
+     * ### Events
+     *
+     * - `Model.beforeDeleteEach` Fired before the batch delete occurs. If stopped the delete
+     *   will be aborted. Receives the event, \Cake\Datasource\ResultSetInterface, and options.
+     * - `Model.afterDeleteEach` Fired after the batch delete has been successful. Receives
+     *   the event, \Cake\Datasource\ResultSetInterface, and options.
+     * - `Model.afterDeleteEachCommit` Fired after the transaction is committed for
+     *   the batch of atomic deletes. Receives the event, \Cake\Datasource\ResultSetInterface, and options.
+     *
+     * ### Example:
+     *
+     * ```
+     * $count = $articles->deleteEach(['published' => true]);
+     * ```
+     *
+     * @param mixed $conditions Conditions to be used, accepts anything Query::where()
+     * can take.
+     * @param array|\ArrayAccess $options An array that will be passed to Query::applyOptions()
+     * @return int Count Returns the affected rows.
+     * @see \Cake\Datasource\RepositoryInterface::delete()
+     */
+    public function deleteEach($conditions, $options = [])
+    {
+        $options += [
+            'finder' => 'all',
+            'options' => [],
+            'atomic' => true,
+            '_primary' => true,
+            'stopOnFailure' => true,
+            'batch' => 1,
+            'checkRules' => true
+        ];
+
+        $invalid = array_intersect(array_keys($options), ['limit', 'offset', 'page']);
+        if (count($invalid) !== 0) {
+            throw new InvalidArgumentException(sprintf('Option conflicts with batch deleting: %s', implode(', ', $invalid)));
+        }
+
+        if((int)$options['batch'] <= 0) {
+            throw new InvalidArgumentException(sprintf('Invalid batch size: %s', (string)$options['batch']));
+        }
+
+        $count = 0;
+
+        $atomic = function () use ($conditions, $options, &$count) {
+
+            $query = $this->find($options['finder'], $options['options'])
+                ->where($conditions)
+                ->applyOptions($options);
+
+            if ($options['batch'] > 0) {
+                $query->limit($options['batch']);
+            }
+
+            $entities = $query->all();
+            if ($entities->count() === 0) {
+                return false;
```

Comment:
@ADmad good eye, but that's a closure method. The false indicates the method should break out of the do/while loop below.